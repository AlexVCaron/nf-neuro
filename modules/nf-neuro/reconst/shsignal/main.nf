process RECONST_SHSIGNAL {
    tag "$meta.id"
    label 'process_single'

    container "scilus/scilpy:2.2.0_cpu"

    input:
        tuple val(meta), path(dwi), path(bval), path(bvec), path(mask) /* optional, value = [] */

    output:
        tuple val(meta), path("*__dwi_sh.nii.gz")   , emit: sh_signal
        path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def max_shell_bvalue = task.ext.max_shell_bvalue ?: 1500
    def b0_thr_extract_b0 = task.ext.b0_thr_extract_b0 ?: 10
    def b0_threshold = task.ext.b0_thr_extract_b0 ? "--b0_threshold $task.ext.b0_thr_extract_b0" : ""
    def shell_to_fit = task.ext.shell_to_fit ?: "\$(cut -d ' ' --output-delimiter=\$'\\n' -f 1- $bval | awk -F' ' '{v=int(\$1)}{if(v<=$max_shell_bvalue|| v<=$b0_thr_extract_b0)print v}' | sort | uniq)"
    def sh_order = task.ext.sh_order ? "--sh_order $task.ext.sh_order" : ""
    def sh_basis = task.ext.sh_basis ? "--sh_basis $task.ext.sh_basis" : ""
    def smoothing = task.ext.smoothing ? "--smooth $task.ext.smoothing" : ""
    def attenuation_only = task.ext.fit_attenuation_only ? "--use_attenuation" : ""

    if ( mask ) args += " --mask $mask"
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_dwi_extract_shell $dwi $bval $bvec $shell_to_fit \
        dwi_sh_shells.nii.gz bval_sh_shells bvec_sh_shells -f

    scil_dwi_to_sh dwi_sh_shells.nii.gz bval_sh_shells bvec_sh_shells \
        ${prefix}__dwi_sh.nii.gz \
        $sh_order $sh_basis $smoothing \
        $attenuation_only $b0_threshold $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_dwi_extract_shell -h
    scil_dwi_to_sh -h

    touch ${prefix}__dwi_sh.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
