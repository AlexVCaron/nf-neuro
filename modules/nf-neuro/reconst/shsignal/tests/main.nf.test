nextflow_process {

    name "Test Process RECONST_SHSIGNAL"
    script "../main.nf"
    process "RECONST_SHSIGNAL"

    tag "modules"
    tag "modules_nfcore"
    tag "reconst"
    tag "reconst/shsignal"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "DWIss1000-dir32.zip", "segmentation.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("fit single shell signal") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "DWIss1000-dir32"
                        segmentation: it.simpleName == "segmentation"
                    }

                ch_dwi = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/brainmask/brainmask.nii.gz")
                    ]
                }
                input[0] = ch_dwi
                    .join(ch_mask)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    niftiMD5SUM(process.out.sh_signal.get(0).get(1), 6),
                    process.out.versions
                ).match() }
            )
        }

    }

    test("fit sh - stub") {

        options "-stub-run"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "DWIss1000-dir32"
                        segmentation: it.simpleName == "segmentation"
                    }

                ch_dwi = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/brainmask/brainmask.nii.gz")
                    ]
                }
                input[0] = ch_dwi
                    .join(ch_mask)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }

    }

}
