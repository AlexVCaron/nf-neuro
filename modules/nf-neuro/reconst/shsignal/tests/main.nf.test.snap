{
    "fit single shell signal": {
        "content": [
            "test__dwi_sh.nii.gz:md5:header,76a9953111f97786bd14a451c61529e7,data,9d8c694457c72375d075ce29838ec91e",
            [
                "versions.yml:md5,b31dee07198d655bc73039659807fcbd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T20:23:58.775324753"
    },
    "fit sh - stub": {
        "content": [
            [
                "versions.yml:md5,b31dee07198d655bc73039659807fcbd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T20:24:12.986333631"
    }
}